File: frontend/todo.js
const inputBox = document.getElementById("input-box");
const button = document.querySelector("button");
const list = document.getElementById("list-container");
const myButton = document.querySelector('#add-task');
const othersTasksLink = document.querySelector('.link-text'); // "View others' tasks" link
const authModal = document.getElementById('auth-modal');
const signupLoginBtn = document.getElementById('signup-login-btn');
const closeBtn = document.querySelector('.close-btn');
const signupForm = document.getElementById('signup-form');
const authSubmitBtn = document.getElementById('auth-submit-btn');
const modalTitle = document.getElementById('modal-title');
const switchToLogin = document.getElementById('switch-to-login');
const logoutBtn = document.getElementById('logout-btn');
const errorMessage = document.getElementById('error-message'); // Error message element
const host_url = 'http://127.0.0.1:8000';

// Function to show the signup/login modal
function showAuthModal() {
    authModal.style.display = 'block';
}

// Check authentication before performing any action
function checkAuthentication(event) {
    if (!sessionStorage.getItem('accessToken')) {
        event.preventDefault(); // Prevent the default action (navigation or form submission)
        showAuthModal(); // Show the signup/login modal
    }
}

// Event listener for "Add" button click
myButton.addEventListener("click", function(event) {
    checkAuthentication(event); // Check if the user is authenticated
    if (sessionStorage.getItem('accessToken')) {
        addTask(); // Call addTask if authenticated
    }
});

// Event listener for "View Others' Tasks" link click
othersTasksLink.addEventListener("click", function(event) {
    checkAuthentication(event); // Check if the user is authenticated
});

// Function to add a new task
async function addTask() {
    // Clear any previous error message
    errorMessage.style.display = 'none';
    errorMessage.textContent = '';

    if (inputBox.value === '') {
        // Show error message below the input field
        errorMessage.textContent = "You must write something!";
        errorMessage.style.display = 'block';
    } else {
        const task = { title: inputBox.value, description: "" }; // Adjust task structure if needed
        try {
            const response = await fetch(`${host_url}/task/tasks/`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${sessionStorage.getItem('accessToken')}`
                },
                body: JSON.stringify(task)
            });

            if (response.ok) {
                showTask(); // Refresh task list after adding new task
                inputBox.value = '';
            } else {
                alert('Error adding task.');
            }
        } catch (error) {
            console.error('Error:', error);
        }
    }
}

// Function to mark a task as completed or not completed
async function completeTask(taskId, currentCompletedStatus) {
    try {
        const updatedStatus = !currentCompletedStatus; // Toggle completed status
        const response = await fetch(`${host_url}/task/tasks/${taskId}/`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${sessionStorage.getItem('accessToken')}`
            },
            body: JSON.stringify({ completed: updatedStatus }) // Send the updated completed status
        });

        if (response.ok) {
            showTask(); // Refresh task list after marking as completed or not completed
        } else {
            alert('Error updating task status.');
        }
    } catch (error) {
        console.error('Error:', error);
    }
}

// Function to show tasks from the server
async function showTask() {
    try {
        const response = await fetch(`${host_url}/task/tasks/`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${sessionStorage.getItem('accessToken')}`
            }
        });

        if (response.ok) {
            const result = await response.json();
            list.innerHTML = ''; // Clear the current list
            if (result.status === 200 && result.data.length > 0) {
                result.data.forEach(task => {
                    let li = document.createElement("li");
                    li.innerHTML = `<strong>${task.title}</strong> ${task.description}`;
                    li.setAttribute('data-completed', task.completed); // Store the completed status in the element
                    list.appendChild(li);

                    let span = document.createElement("span");
                    span.innerHTML = "x";
                    span.setAttribute('data-id', task.id); // Store task id for deletion
                    li.appendChild(span);

                    if (task.completed) {
                        li.classList.add("checked"); // Mark as completed if task is completed
                    }

                    // Add event listener for marking the task as completed or not completed
                    li.addEventListener('click', (e) => {
                        if (e.target !== span) { // Avoid triggering on the delete button
                            const currentCompletedStatus = task.completed;
                            completeTask(task.id, currentCompletedStatus); // Pass current status to toggle
                        }
                    });
                });
            } else {
                list.innerHTML = '<li>No tasks available.</li>';
            }
        } else {
            alert('Error fetching tasks.');
        }
    } catch (error) {
        console.error('Error:', error);
    }
}

list.addEventListener("click", (e) => {
    if (e.target.tagName === "SPAN") {
        const taskId = e.target.getAttribute('data-id');
        deleteTask(taskId); // Call function to delete task
    }
});

// Function to delete a task
async function deleteTask(taskId) {
    try {
        const response = await fetch(`${host_url}/task/tasks/${taskId}/`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${sessionStorage.getItem('accessToken')}`
            }
        });

        if (response.ok) {
            showTask(); // Refresh task list after deletion
        } else {
            alert('Error deleting task.');
        }
    } catch (error) {
        console.error('Error:', error);
    }
}

myButton.addEventListener("click", addTask);

// Show the modal when the signup/login button is clicked
signupLoginBtn.onclick = function () {
    authModal.style.display = 'block';
}

// Close the modal when the close button is clicked
closeBtn.onclick = function () {
    authModal.style.display = 'none';
}

// Close the modal if user clicks outside the modal
window.onclick = function (event) {
    if (event.target == authModal) {
        authModal.style.display = 'none';
    }
}

// Toggle between signup and login
switchToLogin.onclick = function (e) {
    e.preventDefault();
    const nameInput = document.getElementById('name');

    if (modalTitle.textContent === 'Sign Up') {
        // Switch to Login mode
        modalTitle.textContent = 'Login';
        authSubmitBtn.textContent = 'Login';
        nameInput.style.display = 'none';  // Hide the name input for login
        nameInput.removeAttribute('required'); // Remove the required attribute
        switchToLogin.innerHTML = `Don't have an account? <a href="#">Sign Up</a>`;
    } else {
        // Switch to Signup mode
        modalTitle.textContent = 'Sign Up';
        authSubmitBtn.textContent = 'Sign Up';
        nameInput.style.display = 'block';  // Show the name input for signup
        nameInput.setAttribute('required', 'required'); // Add the required attribute
        switchToLogin.innerHTML = `Already have an account? <a href="#">Login</a>`;
    }
}

// Handle signup/login form submission
signupForm.addEventListener('submit', async function (e) {
    e.preventDefault();

    // Get form data
    const name = document.getElementById('name').value;
    const email = document.getElementById('email').value;
    const password = document.getElementById('password').value;

    // Set the URL based on whether we're signing up or logging in
    const url = modalTitle.textContent === 'Sign Up' ? `${host_url}/task/signup-login/` : `${host_url}/task/signup-login/`;
    const isSignup = modalTitle.textContent === 'Sign Up';

    // Prepare payload for signup or login
    const payload = isSignup
        ? { name, email, password }
        : { email, password };

    // Send the request to the API
    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(payload),
        });

        const data = await response.json();

        if (response.ok) {
            // alert(data.message);
            sessionStorage.setItem('accessToken', data.access); // Store access token in session storage
            sessionStorage.setItem('refreshToken', data.refresh); // Store refresh token in session storage
            authModal.style.display = 'none';
            signupLoginBtn.style.display = 'none'; // Hide login/signup button
            logoutBtn.style.display = 'block'; // Show logout button
            showTask(); // Load tasks after login
        } else {
            alert('Error: ' + data.message);
        }
    } catch (error) {
        console.error('Error:', error);
        alert('Something went wrong.');
    }
});

// Handle logout
logoutBtn.onclick = function () {
    sessionStorage.clear(); // Clear session storage
    signupLoginBtn.style.display = 'block'; // Show login/signup button
    logoutBtn.style.display = 'none'; // Hide logout button
    list.innerHTML = ''; // Clear task list
    // alert('Logged out successfully.');
}

// Show tasks on page load if logged in
window.addEventListener("load", function () {
    if (sessionStorage.getItem('accessToken')) {
        signupLoginBtn.style.display = 'none'; // Hide login/signup button
        logoutBtn.style.display = 'block'; // Show logout button
        showTask(); // Load tasks
    }
});

File: frontend/todo.py
import streamlit as st
import requests

# API URL endpoints
SIGNUP_LOGIN_URL = "http://127.0.0.1:8000/task/signup-login/"
TASKS_URL = "http://127.0.0.1:8000/task/tasks/"
OTHERS_TASKS_URL= "http://127.0.0.1:8000/task/tasks/other/"


# Function to handle signup
def signup(name, email, password):
    response = requests.post(SIGNUP_LOGIN_URL, data={"name": name, "email": email, "password": password})
    response.raise_for_status()
    return response.json()

# Function to handle login
def login(email, password):
    payload = {"email": email, "password": password}
    response = requests.post(SIGNUP_LOGIN_URL, data=payload)
    response.raise_for_status()
    return response.json()

# Function to fetch tasks
def get_tasks(access_token):
    headers = {"Authorization": f"Bearer {access_token}"}
    response = requests.get(TASKS_URL, headers=headers)
    response.raise_for_status()
    return response.json()

def create_tasks(title, description, access_token):
    headers = {"Authorization": f"Bearer {access_token}"}
    response = requests.post(TASKS_URL, headers=headers, data={"title": title, "description": description})
    response.raise_for_status()
    return response.json()
    
def get_others_tasks(access_token):
    headers = {"Authorization": f"Bearer {access_token}"}
    response = requests.get(OTHERS_TASKS_URL, headers=headers)
    response.raise_for_status()
    return response.json()


# Create a header with the app name on the left and buttons on the right
def create_header():
    col1, col2 = st.columns([3, 1])
    with col1:
        st.markdown("# To Do List")
    with col2:
        # Apply custom CSS to center the buttons vertically
        st.markdown("""
        <style>
        .centered-content {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100%;
        }
        </style>
        """, unsafe_allow_html=True)

        # Use a container with the centered-content class to align vertically
        with st.container():
            st.markdown("<div class='centered-content'>", unsafe_allow_html=True)
            col3, col4 = st.columns(2)
            with col3:
                signup_clicked = st.button("Signup", key="header_signup_button")
            with col4:
                login_clicked = st.button("Login", key="header_login_button")
            st.markdown("</div>", unsafe_allow_html=True)

            if signup_clicked:
                st.session_state['show_login'] = False
            if login_clicked:
                st.session_state['show_login'] = True

# Main function to control the app flow
def main():
    # Display the header
    if 'access_token' not in st.session_state:
        create_header()

    # Check if the user is logged in
    if 'access_token' in st.session_state:
        # Display sidebar with options
        with st.sidebar:
            st.header("Navigation")
            if st.button("Create Task", key="sidebar_create_task"):
                st.session_state['show_create_task'] = True
                st.session_state['show_your_tasks'] = False
                st.session_state['show_others_tasks'] = False
            if st.button("View Your Tasks", key="sidebar_view_your_tasks"):
                st.session_state['show_create_task'] = False
                st.session_state['show_your_tasks'] = True
                st.session_state['show_others_tasks'] = False
            if st.button("View Others' Tasks", key="sidebar_view_others_tasks"):
                st.session_state['show_create_task'] = False
                st.session_state['show_your_tasks'] = False
                st.session_state['show_others_tasks'] = True

        st.write("## Welcome to your To Do List!")

        # Display the "Your Tasks" page by default
        if 'show_create_task' not in st.session_state:
            st.session_state['show_your_tasks'] = True

        # Display content based on the selected option
        if st.session_state.get('show_create_task'):
            st.write("## Create a Task")
            # Logic to create a task can go here
            try:
                access_token = st.session_state['access_token']
                title = st.text_input("Title", key="task_title")
                description = st.text_input("Description", key="task_description")
                submit_task = st.button("Create", key="submit_task")
                if submit_task:
                    task = create_tasks(title, description, access_token)
                    print(task)
                    # for task in tasks['data']:
                    st.write(f"- {task['data']['title']}: {task['data']['description']}")
                    
            except requests.exceptions.HTTPError as e:
                st.error(f"Failed to create tasks: {e.response.json().get('message', 'Error creating tasks')}")

        elif st.session_state.get('show_your_tasks'):
            st.write("## Your Tasks")
            try:
                access_token = st.session_state['access_token']
                tasks = get_tasks(access_token)
                for task in tasks['data']:
                    col1, col2 = st.columns([3, 1])
                    with col1:
                        st.markdown(f"- {task['title']}: {task['description']}")
                    with col2:
                        st.markdown("""
                            <style>
                            .centered-content {
                                display: flex;
                                justify-content: center;
                                align-items: center;
                                height: 100%;
                            }
                            </style>
                            """, unsafe_allow_html=True)
                        with st.container():
                            st.markdown("<div class='centered-content'>", unsafe_allow_html=True)
                            col3, col4 = st.columns(2)
                            with col3:
                                edit_clicked = st.button("Edit", key=f"edit_button_{task['id']}")
                            with col4:
                                del_clicked = st.button("Delete", key=f"delete_button_{task['id']}")
                            st.markdown("</div>", unsafe_allow_html=True)

                            if edit_clicked:
                                pass
                            if del_clicked:
                                pass

                    # st.write(f"- {task['title']}: {task['description']}")
            except requests.exceptions.HTTPError as e:
                st.error(f"Failed to fetch tasks: {e.response.json().get('message', 'Error fetching tasks')}")

        elif st.session_state.get('show_others_tasks'):
            st.write("## Others' Tasks")
            # Logic to view others' tasks can go here
            try:
                access_token = st.session_state['access_token']
                tasks = get_others_tasks(access_token)
                for task in tasks['data']:
                    st.write(f"- {task['title']}: {task['description']}")
            except requests.exceptions.HTTPError as e:
                st.error(f"Failed to fetch tasks: {e.response.json().get('message', 'Error fetching tasks')}")

    else:
        # Display login/signup forms
        if 'show_login' not in st.session_state:
            st.session_state['show_login'] = False

        # Toggle between signup and login forms
        if st.session_state['show_login']:
            st.write("## Login")
            login_email = st.text_input("Email", key="login_email")
            login_password = st.text_input("Password", type="password", key="login_password")
            login_button = st.button("Login", key="form_login_button")
            
            if login_button:
                try:
                    login_response = login(login_email, login_password)
                    st.success("Login successful!")
                    st.session_state['access_token'] = login_response.get("access")
                    st.session_state['refresh_token'] = login_response.get("refresh")
                    st.session_state['show_tasks'] = True
                    st.session_state['show_your_tasks'] = True  # Set default to show your tasks
                except requests.exceptions.HTTPError as e:
                    st.error(f"Login failed: {e.response.json().get('message', 'Invalid credentials')}")
        else:
            st.write("## Signup")
            name = st.text_input("Name", key="signup_name")
            email = st.text_input("Email", key="signup_email")
            password = st.text_input("Password", type="password", key="signup_password")
            signup_button = st.button("Signup", key="form_signup_button")
            
            if signup_button:
                try:
                    signup_response = signup(name, email, password)
                    st.session_state['access_token'] = signup_response.get("access")
                    st.session_state['refresh_token'] = signup_response.get("refresh")
                    st.success("Signup successful! You can now log in.")
                    st.session_state['show_login'] = True  # Switch to login form
                    st.session_state['show_tasks'] = True
                    st.session_state['show_your_tasks'] = True  # Set default to show your tasks
                except requests.exceptions.HTTPError as e:
                    st.error(f"Signup failed: {e.response.json().get('message', 'Unknown error')}")

if __name__ == "__main__":
    main()
